src/kudu/consensus/consensus_peers.cc

```diff
@@ -243,6 +243,7 @@ void Peer::SendNextRequest(bool even_if_queue_empty) {
       request_.committed_index() : kMinimumOpIdIndex;
 
   if (PREDICT_FALSE(!s.ok())) {
+    failed_attempts_++;
```

Comment:
I think there is a problem here. Even heartbeats go through the same path. So, a lagging peer will not receive heartbeats and if that particular peer is not removed from the ring soon (by external automation), it could end up starting pre-elections and destabilizing the ring. I wonder why upstream kudu did it this way. Maybe, in a future diff we could explore a possibility of degenerating this case of "lagging peer will never be able to catch up because logs are GCd" into a heartbeat-with-status only message to prevent the lagging peer from starting elections.

Reply:
Yes, since this error is at RequestForPeer, there is no way to make a heartbeat go through in current state. The current change only impacts how many times that is attempted. Either the approach that you are suggesting or somehow make a no-ops based heartbeat go through just to maintain authority.